RFC: drivers: stm32: SPI: Check that SPI buffers are in a nocache region

dgastonochoa opened the pull request:
This is a proposal for adding a memory region to the STM32H7 SoCs to be used by  DMA peripherals.

In H7, DMA cannot access certain memory regions. A peripheral using DMA will not work correctly if the provided read/write buffers are in the wrong memory regions. This can lead to confusing problems, e.g. if SPI is used with DMA but with buffers at the wrong addresses, SPI will generate its clock bu it will not send anything. Furthermore, if the buffers are in a valid region but the latter is not configured as nocache, SPI will still generate its clock but send all zeroes, or even worse, it will send the data that was already present in that memory region (for instance, if it was initialised at startup time) but it won't send new values written to such region.

This PR reclares a memory region in SRAM3, 2 or 1 (depending on which one is available) and sets them as nocache. This is equivalent to what is already done for Ethernet in H7.

This intents to be the first of 2 PRs: if this gets approved, the plan is to upload another PR to verify, within the stm32 SPI driver, that the given send/receive buffers are within the correct regions, or return an error otherwise, to warn the user.

How to test
===========
Below there is a file which is a Zephyr application called 'dma-test' that can be placed in zephyr/samples/basic, then built and run to test the PR. It has been tested on a nucleo-h743zi board. It sends some data over SPI (with DMA enabled), then it modifies that data, and then it sends it again to verify DMA got the changes.

It also contains a test to see what happens if a buffer, which is placed in an invalid memory region, is sent. For now, SPI does not return an error. It tries to send it, but no data will be actually sent as explained above.

[dma-test.tar.gz](https://github.com/zephyrproject-rtos/zephyr/files/11451569/dma-test.tar.gz)

You can build it with: `west build -b nucleo_h743zi zephyr/samples/basic/dma-test`

Here is a logic analyzer capture:

![asd](https://github.com/zephyrproject-rtos/zephyr/assets/127399215/6a7596fe-d6e6-46e6-a371-e392539dd222)

Above you can see how the first 2 data chunks send the expected data but the last one sends garbage, as the buffer is in the wrong regions.

erwango commented:
Thanks for this proposal, this problem has been identified since some time, see https://github.com/zephyrproject-rtos/zephyr/issues/36471 and https://github.com/zephyrproject-rtos/zephyr/issues/57220.
Note that the solution you proposed was evoked in #36471 with following evaluation: "this is a significant constraint to map user buffers in this NoCache area, else memcopy are also needed."

I suggest to follow https://github.com/zephyrproject-rtos/zephyr/issues/57220 going forward.

This being said, I'll try to review the proposal for the sake of discussion

dgastonochoa commented:
Hi @erwango, would this change make sense as short-term solution until #57220 is carried on?

erwango commented on soc/arm/st_stm32/stm32h7/mpu_regions.c:
```diff
@@ -21,18 +23,36 @@ static const struct arm_mpu_region mpu_regions[] = {
+					 REGION_RAM_NOCACHE_ATTR(REGION_4K)),
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(sram3), okay) */
 	/* DT-defined regions */
 	LINKER_DT_REGION_MPU(ARM_MPU_REGION_INIT)
```
This macro should always be called, it can't be conditional.

dgastonochoa replied:
No longer applicable as a different implementation has been requested and applied.

dgastonochoa commented:
I see, please let me double check if I have understood it correctly:

I would need to enable DMA and declare a memory region as nocache in a DTS overlay file, as `tests/drivers/adc/adc_dma/boards/nucleo_h743zi.overlay` does, e.g.:
```
/* ADC driver expects a buffer in a non-cachable memory region */
&sram4 {
	zephyr,memory-region-mpu = "RAM_NOCACHE";
};
```
Then, in my app. source file, I would declare the buffer to be used with DMA to be placed in "SRAM4.dma" (SRAM4 or whatever region I declare as nocache). This would all be application-level, so it must not be in the PR.

The PR should do something similar to `drivers/adc/adc_stm32.c::address_in_non_cacheable_sram` but in SPI (which accomplishes to what my planned second PR was going to do). Is this correct? 

If so, is there any good common place to put the `address_in_non_cacheable_sram` function, so it's accessible for all drivers? I would like to avoid copying it to the SPI driver.

erwango commented:
> If so, is there any good common place to put the address_in_non_cacheable_sram function, so it's accessible for all drivers? I would like to avoid copying it to the SPI driver.

Maybe something like `soc/arm/st_stm32/stm32h7/mpu_regions.c`?

dgastonochoa commented:
I haven't found any place that seems right to put this function (`mpu_regions.c` has not a public header). Also, modifying the ADC driver to take it from a common place would require further testing. Thus, I have placed the new logic in the SPI driver directly.

dgastonochoa commented:
@erwango my understanding from your last comment is that this is the required change, is this correct? Thanks.

erwango reviewed (changes requested):
This is going in the right direction.
Also, I'd like @FRASTM to confirm this is in line with latest changes in STM32H7 SPI area.

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -723,6 +799,13 @@ static int transceive_dma(const struct device *dev,
 		return -ENOTSUP;
 	}
 
+#ifdef CONFIG_SOC_SERIES_STM32H7X
+	if ((tx_bufs != NULL && !all_bufs_in_valid_dma_ram(tx_bufs)) ||
+		(rx_bufs != NULL && !all_bufs_in_valid_dma_ram(rx_bufs))) {
+		return -EFAULT;
+	}
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
```
Oups, my bad. I actually checked the function name, but somehow misread.
That's fine, indeed!

dgastonochoa commented:
Thanks for the review; It still needs a second approval :)

teburd commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -679,6 +679,82 @@ static int transceive(const struct device *dev,
 #ifdef CONFIG_SPI_STM32_DMA
 
+#ifdef CONFIG_SOC_SERIES_STM32H7X
+#define STM32_SPI_GET_MEM_REGION(node_id)                                       \
```
This seems more broadly useful, can we move it to to one of the devicetree related headers where other DT macros live?

dgastonochoa replied:
I'm doubtful about putting it in `include/zephyr/devicetree/dma.h` or create a new file `include/zephyr/devicetree/mem.h`

teburd replied:
Second option makes good sense to me, its about inspecting pointers to determine if they lie within devicetree defined memory regions?

dgastonochoa replied:
I have taken a look at the rest of the code and the second option seems inconsistent with what Zephyr does. For instance, the driver `gpio.h` declares a `struct gpio_dt_spec` and provides a macro `GPIO_DT_SPEC_GET_BY_IDX` to initialize it. To follow the same pattern, I would put this new struct and macro in the memory driver header. The problem is that there are no memory drivers in Zephyr. The next most similar option seems to me to use `zephyr/include/zephyr/drivers/dma.h`, as this is very related to DMA. That will work for the struct and macro, but what to do with the functions? They are generic, don't belong to a particular driver, soc or board. They only place in which they could fit is somewhere under `subsys`. There is nothing related to memory there yet, so I would need to create a new directory etc. to take this option. 

Another option is, as explained above, to put the struct and the macro in the DMA driver header file and add there the functions as well as `static inline`, header-defined functions. I'm not a big fan of header-defined functions, however, in this case, it might be the only reasonable option.

Which option do you find better?

dgastonochoa replied:
I think I'm going to go with the last option, seems the best one given the circumstances.

dgastonochoa replied:
In the end I have opted for the DMA option as it seemed simpler and I didn't need to modify kernel files.

teburd reviewed (commented):
I like it, and it sort of ties into what I was hoping to see more broadly with DT and buffers, which means I'd really love to see this in a more general place in Zephyr

teburd commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -679,6 +679,33 @@ static int transceive(const struct device *dev,
+ * with DMA.
+ *
+ */
+static bool all_bufs_in_valid_dma_ram(const struct spi_buf_set *bufs)
```
probably should go in spi.h

spi_buf_set_in_nocache(const struct spi_buf_set *bufs)

dgastonochoa replied:
I see what you mean, will do.

dgastonochoa replied:
Bear in mind though that moving this to the `spi.h` header will involve to add an `#include <drivers/dma.h` to it as well, as the body of the function uses the `dma_mem_region` type. I have considered moving all this to either `devicetree.h` or some header in `include/zephyr/devicetree/*` but no structs/functions are declared there, so it doesn't seem consistent.

teburd commented on include/zephyr/drivers/dma.h:
```diff
@@ -256,6 +258,50 @@ struct dma_context {
+ * a string that specifies such region's properties.
+ *
+ */
+struct dma_mem_region {
```
I still think this would be better represented by a dt_mem_region, its not just dma is it? In device tree memory regions have their own subtree I believe

dgastonochoa replied:
Yes, yesterday night I was still thinking about it and I didn't create the `include/zephyr/devicetree/mem.h` header (mentioned in the previous comment) to avoid having a static inline function in a header file, but I ended up adding it here. As you say, in summary, I will move the SPI function above to its header file so other SPI drivers can use it, and I will move all the code in the `dma.h` header to a new file `include/zephyr/devicetree/mem.h`

dgastonochoa replied:
Since it is declared in `dma.h`, shouldn't it be `dma_dt_mem_region`? As `gpio_dt_spec` does.

teburd replied:
I'm saying the binding has nothing to do with dma, https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/base/zephyr%2Cmemory-region.yaml it's a general purpose memory region and should be DT related

dgastonochoa replied:
All moved to `include/zephyr/devicetree/mem.h`, as discussed before.

teburd commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,111 @@
+#define MEM_NOCACHE_ATTR "RAM_NOCACHE"
+
+/**
+ * @brief Container for MMIO SRAM information specified in devicetree
```
Could be other memory regions, sdram, fram, etc, I wouldn't predispose this to mmio sram bindings as the base binding is the base memory-region one

dgastonochoa replied:
Done

teburd commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,111 @@
+static inline bool dt_buf_in_nocache(
+			const uint8_t *buf,
+			size_t size,
+			const struct dt_mem_region mem_regions[],
```
Is there a way we could avoid this parameter by walking DT for all memory regions perhaps and coalescing them into an iterable section of dt_mem_region structs?

teburd reviewed (dismissed):
Looks good, would love to be able to fire and forget the call to check all memory regions rather than having to supply them but I get that might require some dt python generator changes

dgastonochoa commented:
The Clang build failed; fixed now.

carlocaione commented:
This is something very similar to what I did a while ago with this https://docs.zephyrproject.org/latest/kernel/memory_management/shared_multi_heap.html (see also the sample / test at https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/kernel/mem_heap/shared_multi_heap ).

carlocaione commented:
@teburd @dgastonochoa so, thinking out loud about this problem and how the `shared_multi_heap` can be used to solve this kind of problems and many others related to #57220.

What about:
1. Implementing something similar to `DT_GET_MEM_REGION` from this PR so that `shared_multi_heap_region` can be easily populated
2. Adding a new API function to the `shared_multi_heap` similar to what `dt_buf_in_nocache` is doing in this PR
3. (bonus) Change name to something more general?

carlocaione reviewed (dismissed):
Blocking this for now because I want to discuss a more general approach to the same problem.

carlocaione commented:
> I'm more inclined to approve and merge this change as it is and create a new issue/RFC with your suggestions if required @carlocaione.
> 
> Besides, even in case the changes you proposed are applied, in my opinion it still makes sense to leave this new functionality where it is, strictly related to devicetree, and then use it from `shared_multi_heap`.
> 
> Do you agree @erwango and @teburd?

The problem is that this solution is not scalable at all.

1. You have one single attribute (what if you need to add more attributes in the future? or you want to check multiple attributes at the same time?)
2. Related to the 1st point a buffer that is non cacheable is not automatically DMA-able, so you need right away one more attribute (and you cannot even use `zephyr,memory-region-mpu` for that)
3. This is not considering MMU-enabled platforms at all where the virtual address is different than the physical address retrieved from DT
4. `spi_buf_set_in_nocache` is going to be replicated in each subsystem then? I2C, I3C, DMA, etc...

dgastonochoa commented:
> You have one single attribute (what if you need to add more attributes in the future? or you want to check multiple attributes at the same time?)

I don't see why. `region_mpu` is a string that can contain more than 1 attribute. See
```
enum:
      - "RAM"
      - "RAM_NOCACHE"
      - "FLASH"
      - "PPB"
      - "IO"
```
In `zephyr,memory-region.yaml`. If more attributes are needed they can be checked as well. Maybe you mean other attributes that are not in that list? In such case the problem would rather be adding such attributes to the list above, right? Or are there attributes that must not be in devicetree?


---

> Related to the 1st point a buffer that is non cacheable is not automatically DMA-able, so you need right away one more attribute (and you cannot even use zephyr,memory-region-mpu for that)

This PR aims to provide a way to check if a buffer is in a nocache region, as the title indicates. What you mention about the memory regions is true, but out of this RFC's scope as I proposed it (in the second description, after receiving the first feedback). I add this check to the SPI driver because it still works in the sense that being in a nocache area is a necessary-but-not-sufficient condition for DMA to work in stm32h7 platforms (to which this RFC is limited to).

I see that the second description can be misleading, as it mentions "not in a valid memory region" which can be interpreted taking into account the location of the memory and not only if it is nocache. However, immediatly after I write:

> Notice that the problem of detecting if the buffers are in a memory area accessible by DMA (regardless it is nocache or not) remains.

So hopefully that clarifies it a bit. I will change the description to reflect what was my intention for this RFC: to only check if the buffers are in a nocache region.

---

> This is not considering MMU-enabled platforms at all where the virtual address is different than the physical address retrieved from DT

True, it just only aims to provide a mechanism to determine if a buffer is in a nocache region. In particular, it adds a test tp the SPI dma-transceive for stm32h7 platforms only.

> spi_buf_set_in_nocache is going to be replicated in each subsystem then? I2C, I3C, DMA, etc...

`spi_buf_set_in_nocache` exists because of the peculiarity of SPI by which it receives a set of buffers as a parameter instead of a single buffer. Thus, it will not be replicated; other peripherals would call `dt_buf_in_nocache` directly.  I don't know if other peripherals receive a set of buffers as arguments, but in that case yes: they will need to replicate this functionality.

---

In summary, I agree with most what you say, specially with the memory region problem. But I insist, my intention was to keep this PR's scope as limited as possible, and apply that change and others (e.g. the one proposed by @teburd above) in further PRs, as per the Zephyr contributor guidelines:

> The Zephyr project encourages contributors to submit changes as smaller pull requests.

carlocaione commented:
> > You have one single attribute (what if you need to add more attributes in the future? or you want to check multiple attributes at the same time?)
> 
> I don't see why. `region_mpu` is a string that can contain more than 1 attribute. See
> 
> ```
> enum:
>       - "RAM"
>       - "RAM_NOCACHE"
>       - "FLASH"
>       - "PPB"
>       - "IO"
> ```
> 
> In `zephyr,memory-region.yaml`. If more attributes are needed they can be checked as well. Maybe you mean other attributes that are not in that list? In such case the problem would rather be adding such attributes to the list above, right? Or are there attributes that must not be in devicetree?

Right, but this an even bigger problem because those attributes are exclusively used by ARM with MPU, so you are basically introducing a feature that is exclusively usable on ARM. Those attributes are not usable on any other architecture, especially not with MMU. An additional layer is needed to decouple those attributes from the architecture. See for example what was done in: https://github.com/zephyrproject-rtos/zephyr/blob/9e0e2be7651cbabc142e6e9e706c91f638defc75/include/zephyr/multi_heap/shared_multi_heap.h#L69-L78 ).
You need some kind of translation layer from DT arch-specific attributes to general attributes.

About adding new attributes that is not really feasible. The `zephyr,memory-region-mpu` is used to tweak the MPU regions, if you try to add any attribute that is not meaningful for the MPU (so again ARM specific) you will get an error at compile time.

> `spi_buf_set_in_nocache` exists because of the peculiarity of SPI by which it receives a set of buffers as a parameter instead of a single buffer. Thus, it will not be replicated; other peripherals would call `dt_buf_in_nocache` directly. I don't know if other peripherals receive a set of buffers as arguments, but in that case yes: they will need to replicate this functionality.

The problem to check whether a buffer is allocated in cacheable or not-cacheable memory is a general problem, not really specific to SPI or this device in particular.

> In summary, I agree with most what you say, specially with the memory region problem. But I insist, my intention was to keep this PR's scope as limited as possible, and apply that change and others (e.g. the one proposed by @teburd above) in further PRs, as per the Zephyr contributor guidelines:
> 
> > The Zephyr project encourages contributors to submit changes as smaller pull requests.

I understand that, but IMO this is a very limited solution to a very general (and complicated) problem.

dgastonochoa commented:
I would like to have an agreement about this before doing anything else.

Please @erwango and @teburd could you confirm whether or not you agree with modifying this PR as per the above suggestion, or if you prefer to merge this change and create other PRs to address its drawbacks? 

Let me quote mentioned suggestion here for simplicity:

> What about:
> 
>     1. Implementing something similar to `DT_GET_MEM_REGION` from this PR so that `shared_multi_heap_region` can be easily populated.
> 
>     2. Adding a new API function to the `shared_multi_heap` API similar to what `dt_buf_in_nocache` is doing in this PR to check the properties of the memory region.
> 
>     3. Adding a `no_allocate` flag to exclude some memory region from populate the heap so the user can statically allocate buffers from that regions and still use the new function introduced at point (2).
> 
>     4. (bonus) Change name to something more general?

Thanks

hubertmis reviewed (commented):
It looks current shape of this RFC assumes that SPI driver users are aware of the platform for which the application is being built. The SPI driver user (like an SPI display driver) is forced to declare non-cacheable buffers for selected platforms, and potentially cacheable with others (ifdefs?). It breaks compatibility with existing drivers using SPI and implies that creating SPI-based drivers will be more complicated to maintain compatibility. I'm not sure if that's the intended direction of shaping the SPI driver API.

dgastonochoa commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -733,6 +733,13 @@ static int transceive_dma(const struct device *dev,
+#ifdef CONFIG_SOC_SERIES_STM32H7X
+	if ((tx_bufs != NULL && !spi_buf_set_in_nocache(tx_bufs)) ||
+		(rx_bufs != NULL && !spi_buf_set_in_nocache(rx_bufs))) {
+		return -EFAULT;
```
I wouldn't say exactly not compatible. In these 2 cases, if STM32H7 SPI is used and DMA is enabled, these functions will silently fail, so I still think it's better to return an error code.

Benni77 replied:
> Of course, I prefer returning an error code instead of silently failing. But I was thinking if we could aim for something like: ...

Hi, I am currently facing the task of installing BLE on the STM32H747 via hci_spi. As long as I use `CONFIG_SPI_STM32=y` or `CONFIG_SPI_STM32_INTERRUPT=y`, it works fine. But with `CONFIG_SPI_STM32_DMA=y` I get `error -14` (EFAULT), from exactly this line.

Am I right in assuming that the hci_spi driver - at least on the STM32H7 - currently has no DMA support? Would the buffers in drivers/bluetooth/hci/spi.c have to be provided with `__attribute__((__section__(".nocache")))` on the STM32H7?

Any suggestions on how I should proceed to use hci_spi with DMA?
Thanks for the help!

dgastonochoa replied:
> Am I right in assuming that the hci_spi driver - at least on the STM32H7 - currently has no DMA support? Would the buffers in drivers/bluetooth/hci/spi.c have to be provided with __attribute__((__section__(".nocache"))) on the STM32H7?

Hi, if you want to use SPI in H7 with DMA enabled, the buffers provided to the SPI transceive function must be in a nocache memory region. The preferred way to do this explained in this document: `${ZEPHYR_HOME}/doc/services/mem_mgmt/index.rst`

In summary, you can declare a memory region as nocache in devicetree, then use `__attribute__((section(...)))` to place the required buffer in such region.

erwango replied:
@Benni77 Yes, it makes sense. We need to be able to return a meaningful error in such case.

dgastonochoa commented:
> It looks current shape of this RFC assumes that SPI driver users are aware of the platform for which the application is being built. 

With or without this RFC, at the current status of Zephyr, the user needs to be aware of which memory banks the target platform's DMA can access and under which conditions. This RFC aims to return an error code when something goes wrong, instead of let SPI fail silently and the user needing to spend a considerable amount of time debugging why.

> The SPI driver user (like an SPI display driver) is forced to declare non-cacheable buffers for selected platforms, and potentially cacheable with others (ifdefs?).

If the user does not declare those buffers in non-cacheable memory, the SPI transaction will not work anyway (assuming DMA is enabled). Furthermore, it will silently fail as explained in the RFC description.

>  It breaks compatibility with existing drivers using SPI and implies that creating SPI-based drivers will be more complicated to maintain compatibility.

I don't see how, could you explain it please?

dgastonochoa commented on include/zephyr/drivers/spi.h:
```diff
@@ -1306,6 +1307,35 @@ static inline int spi_release_dt(const struct spi_dt_spec *spec)
 }
 
+/**
+ * @brief Checks that all buffers in `bufs` are in nocache memory region. This
+ * should be used in platforms where DMA cannot access cache memory, and so any
+ * buffer used for DMA operations must be in a nocache region.
```
I agree; that would require a new SPI private header file common to all SPI drivers, as this is "made public" here because it was deemed of general use in changes requested above.

hubertmis replied:
Maybe this one function and one macro would justify creating a `drivers/spi/spi.h` file to move private parts there. Personally, I would create it even if there was only one function to clearly separate "public" from "private". But I believe that's a personal taste :)

carlocaione commented:
> > It looks current shape of this RFC assumes that SPI driver users are aware of the platform for which the application is being built.
> 
> With or without this RFC, at the current status of Zephyr, the user needs to be aware of which memory banks the target platform's DMA can access and under which conditions. This RFC aims to return an error code when something goes wrong, instead of let SPI fail silently and the user needing to spend a considerable amount of time debugging why.

BTW, this is exactly why I'm trying to push this https://github.com/zephyrproject-rtos/zephyr/pull/60049

dgastonochoa commented:
@carlocaione This RFC should now wait for your #60049 to be merged and then use those functions to perform the check I originally wanted to do, am I right?

carlocaione commented:
> @carlocaione This RFC should now wait for your #60049 to be merged and then use those functions to perform the check I originally wanted to do, am I right?

Not really, we can merge this and fix this later. Withdrawing my NACK.

hubertmis commented:
> This RFC aims to return an error code when something goes wrong, instead of let SPI fail silently and the user needing to spend a considerable amount of time debugging why.

Yes, I agree it's better. And I think that's a very good temporary solution. But ultimately I think we should have a solution working fine with the existing samples and drivers.

> I don't see how, could you explain it please?

Sure!

I have a project using a display driver:

https://github.com/zephyrproject-rtos/zephyr/blob/77aafa51a1b5c643d95938b37fd4c9b60baa2516/drivers/misc/ft8xx/ft8xx_drv.c#L112

I'm afraid if I tried to port this project to STM32H7 it wouldn't work, because this driver allocates buffers on a stack that would be cacheable:

https://github.com/zephyrproject-rtos/zephyr/blob/77aafa51a1b5c643d95938b37fd4c9b60baa2516/drivers/misc/ft8xx/ft8xx_drv.c#L90

To make it work with STM32H7 I would need to modify the FT8xx driver to something like

```
#if defined CONFIG_SOC_STM32H7
MACRO_CREATING_NOCACHE_BUFFER(addr_buf, ADDR_SIZE);
#elif defined OTHER_POTENTIAL_PLATFORM_SPECIFIC_CONFIG_THAT_SHOULD_NOT_BE_HERE
...
#else
uint8_t addr_buf[ADDR_SIZE];
#endif
```

I would prefer to modify the STM32H7 driver to be able to accept any buffer, and copy it to non-cacheable memory if the one passed is cacheable. Similarly to these drivers:

https://github.com/zephyrproject-rtos/zephyr/blob/77aafa51a1b5c643d95938b37fd4c9b60baa2516/drivers/serial/uart_nrfx_uarte.c#L1528

https://github.com/zephyrproject-rtos/zephyr/blob/77aafa51a1b5c643d95938b37fd4c9b60baa2516/drivers/spi/spi_nrfx_spim.c#L308

hubertmis commented:
> I respect you prefer that solution, but whether or not it's the best one is an object of discussion, see #57602 as mentioned in the comments above.
> 
> This RFC aims to only detect if a buffer is in a nocache memory region (which is one of the solutions described in the PR above).

This clarifies that in this PR it is intended that existing drivers using SPI won't work with STM32H7 if DMA is enabled. That's fair enough for me and I agree we can discuss in #57602 how to overcome this limitation in a generic way.

> Arguably, you would need to do that in any platform in which DMA cannot access any memory region. Besides, you would do something like `#if STM32H7 && DMA_ENABLED`.

Exactly. And I think we should avoid dozens of lines of platform-specific code in generic drivers. TBH, I think we should avoid any platform-specific code in generic drivers.

dgastonochoa commented:
Hi all, as per [the Zephyr Contributor Expectations](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-review-escalation), sending a kind reminder to this PR's reviewers:

@erwango @ABOSTM @FRASTM @carlescufi @tbursztyka @nashif @galak @MaureenHelm 

Thanks

dgastonochoa commented:
I re-tested this PR a found a problem: two functions were operating with reversed logic:

- `dt_buf_in_nocache` was returning `false` in case a buffer was in a nocache region, and `true` otherwise. This should be the opposite.
- `spi_buf_set_in_nocache` should do `if (!dt_buf_in_nocache(...)) ... ` (notice the negation '!') and it was doing the opposite.

Due to the "double negation" above, the tests explained before passed, but it is fixed now and the tests still pass. 

Naturally, I will re-test it again before merging if this PR is finally accepted, but it's worth reviewers pay special attention to verify that the logic is not "reversed" with regards to the functions' names, just in case I made a mistake.

erwango commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,111 @@
+ *		};
+ *
+ */
+#define DT_GET_MEM_REGION(node_id)                                              \
+	{                                                                           \
+		.start = (const uint8_t *)DT_REG_ADDR(node_id),                         \
+		.end =                                                                  \
+			(const uint8_t *)(DT_REG_ADDR(node_id) + DT_REG_SIZE(node_id)),     \
+		.region_mpu = COND_CODE_1(                                              \
+			DT_NODE_HAS_PROP(node_id, zephyr_memory_region_mpu),                \
+			(DT_PROP(node_id, zephyr_memory_region_mpu)),                       \
+			(NULL)                                                              \
+		)                                                                       \
+	},
```
I'm not devicetree maintainer, but usually request is to some dedicated tests for each new DT macro.
See tests/lib/devicetree/api/src/main.c.

dgastonochoa replied:
I have added tests for `DT_GET_MEM_REGION` and for `dt_buf_in_nocache`

erwango reviewed (changes requested):
Nice work. Looks OK to me but new DT macros should come with dedicated tests

dgastonochoa committed:
drivers: stm32: SPI: Check that SPI buffers are in a nocache region

DMA only works with non-cached memory regions in H7. Check them
and return an error if they don't match this condition.

Signed-off-by: Daniel Gaston Ochoa <dgastonochoa@gmail.com>

carlocaione commented:
> `dt_mem_region` should be adapted to  something like `struct vnd_memory_binding`
> in `tests/lib/devicetree/api/src/main.c`.

That is not strictly needed for what you want to do but it is fine I guess.

> `DT_GET_MEM_REGION(node_id)` should be replaced modified to do
> something like `TEST_FUNC` in `tests/lib/devicetree/api/src/main.c`

Correct.

>  The rest of the functions would remain more or less the same.

Yes.

> The tests implemented in [Move to and extend `zephyr,memory-attr` #60049](https://github.com/zephyrproject-rtos/zephyr/pull/60049)
> should be adapted to these functions (they will no longer need to declare them)

Mmh no, those tests are only needed when you are introducing a new DT helper / macro but in your case you are not introducing anything new, so it is not needed to have tests.

dgastonochoa commented:
@carlocaione  I have started doing the changes mentioned above and I think there is a problem. `DT_MEMORY_ATTR_APPLY` relies on the existence of macro definitions such as `REGION_RAM_ATTR`. As far as I understand, these macros need to have a very specific name, as they need to contain the `REGION_` prefix and the `_ATTR` suffix. The middle part needs to match a name in the enum `zephyr,memory-attr`, defined in `dts/bindings/base/zephyr,memory-attr.yaml`. Is this correct?

If so, the problem I have found is that such macros happen to collision with `REGION_RAM_ATTR(size)` and `REGION_RAM_NOCACHE_ATTR(size)`, defined in `include/zephyr/arch/arm/aarch32/mpu/arm_mpu_v7m.h`.

I'm preparing a PR to fix this problem, by adding a prefix to the macros used by `DT_MEMORY_ATTR_APPLY`. Am I missing something? Thanks

carlocaione commented:
> I'm preparing a PR to fix this problem, by adding a prefix to the macros used by `DT_MEMORY_ATTR_APPLY`. Am I missing something? Thanks

Why do you need to use `DT_MEMORY_ATTR_APPLY` at all? That macro is only useful if you want to create an MPU region and this is not your case. You can simply use `DT_MEMORY_ATTR_FOREACH_NODE` instead to create the struct array as you want.

dgastonochoa commented:
> > I'm preparing a PR to fix this problem, by adding a prefix to the macros used by `DT_MEMORY_ATTR_APPLY`. Am I missing something? Thanks
> 
> Why do you need to use `DT_MEMORY_ATTR_APPLY` at all? That macro is only useful if you want to create an MPU region and this is not your case. You can simply use `DT_MEMORY_ATTR_FOREACH_NODE` instead to create the struct array as you want.

Ah, yes, sorry. I misunderstood it.

carlocaione commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,111 @@
+struct dt_mem_region {
+	const uint8_t *start;
+	const uint8_t *end;
+	const char *region_mpu;
```
maybe call it `region_attr`

dgastonochoa replied:
I have replaced it by a `bool` as per your comment below.

carlocaione commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,111 @@
+			(buf >= mem_reg->start) && ((buf + size) <= mem_reg->end);
+		if (buf_within_bounds &&
+			reg_mpu != NULL &&
+			strncmp(reg_mpu, MEM_NOCACHE_ATTR, sizeof(MEM_NOCACHE_ATTR)) == 0) {
```
instead of doing `strncmp` here you can filter directly when doing `DT_GET_MEM_REGION`.

See https://github.com/zephyrproject-rtos/zephyr/blob/4d56a318f162b758daf04525143380da373e1814/tests/lib/devicetree/api/src/main.c#L2719-L2721

dgastonochoa replied:
Done.

carlocaione commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,111 @@
+		DT_MEMORY_ATTR_FOREACH_NODE(DT_GET_MEM_REGION)
+	};
+
+	size_t mem_regions_cnt = sizeof(mem_regions) / sizeof(*mem_regions);
```
`ARRAY_SIZE`

dgastonochoa replied:
Done

carlocaione commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,111 @@
+		const char *reg_mpu = mem_reg->region_mpu;
+
+		const bool buf_within_bounds =
+			(buf >= mem_reg->start) && ((buf + size) <= mem_reg->end);
```
are you sure about the `<=`?

dgastonochoa replied:
It is indeed `<=` because `mem_reg->end` represents the last memory address in the memory region, which by definition is within the bounds of such region. However, what was wrong is this:
```
.end = (const uintptr_t)(DT_REG_ADDR(node_id) + DT_REG_SIZE(node_id))
```
which should be
```
.end = (const uintptr_t)(DT_REG_ADDR(node_id) + DT_REG_SIZE(node_id)) - 1
```
Right?

dgastonochoa replied:
For the same reason, `(buf + len_bytes)` should be `(buf + len_bytes - 1)` because that will be the last address of the buffer. This is equivalent to do `(buf + len_bytes) <  mem_reg->end` but that is more confusing I think.

dgastonochoa commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,111 @@
+ */
+struct dt_mem_region {
+	const uint8_t *start;
+	const uint8_t *end;
```
Yes, done.

dgastonochoa commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,111 @@
+
+/*
+ * Copyright (c) 2020, Linaro Ltd.
+ * Copyright (c) 2020 Nordic Semiconductor
```
Done

carlocaione commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,107 @@
+ *		};
+ *
+ */
+#define DT_GET_MEM_REGION(node_id)                                              \
+	{                                                                           \
+		.start = (const uintptr_t)DT_REG_ADDR(node_id),                         \
+		.end =                                                                  \
+			(const uintptr_t)(DT_REG_ADDR(node_id) + DT_REG_SIZE(node_id)) - 1, \
+		.isnocache = _DT_IS_NOCACHE_MEM_REGION(node_id)                         \
+	},
```
why a generic helper `DT_GET_MEM_REGION` is only dealing with the `RAM_NOCACHE` attribute?

dgastonochoa replied:
Same explanation as the above comment.

carlocaione commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,107 @@
+ */
+static inline bool dt_buf_in_nocache(uintptr_t buf, size_t len_bytes)
+{
+	const struct dt_mem_region mem_regions[] = {
+		DT_MEMORY_ATTR_FOREACH_NODE(DT_GET_MEM_REGION)
+	};
```
hard no. You are actually allocating RODATA by only including an header file whether you use it or not.

dgastonochoa replied:
Again, this was in the SPI driver source file only until I got requested to move to a generic place.

carlocaione replied:
well, an header is definitely not a proper place for this.

carlocaione commented on tests/lib/devicetree/api/src/main.c:
This is a nice work but (sorry) totally useless at the moment. We can have generic helpers to deal with memory attributes but that is a totally different work. If you want to deal with SPI it is fine as long as you keep everything contained in the SPI layer.

dgastonochoa replied:
> If you want to deal with SPI it is fine as long as you keep everything contained in the SPI layer.

That is exactly what I did in my first commit. This is already going to much out of scope for me. Can you please @erwango , @teburd and @carlocaione agree whether you want this to be generic or self-contained in the SPI driver? I repeat, my only purpose for this PR is to protect the SPI driver `transceive_dma` function to return an error in case of a buffer is in a no-nocache memory area, which I think it's a good idea given that otherwise the SPI driver behaviour is very confusing. I never intended to make this a generic solution.

carlocaione replied:
The rationale here is simple:
- If you want to introduce helpers in `include/zephyr/devicetree/mem.h` these helpers must be generic and deal with all kind of memory attributes, not just SPI and not just CACHE / NOCACHE. This is a pretty big and complex work.
- If you are only interested in SPI + CACHE / NOCACHE, then everything must be self-contained in the SPI layer and possibly header file.

dgastonochoa replied:
> If you are only interested in SPI + CACHE / NOCACHE, then everything must be self-contained in the SPI layer and possibly header file.

I am only interested in SPI + CACHE / NOCACHE. Moreover, I'm interested in STM32 SPI only, so please can we agree to go back to the scope I proposed initially and modify **only** `spi_ll_stm32.c` to perform this check? Not even the header file, as this will require to allocate the memory region array in it as you mentioned. I want to restrict it to solely STM32 SPI.This is the approach that was adopted in #52965.

erwango replied:
@dgastonochoa If there's a solution fitting between @carlocaione and @teburd that will be fine with me.

dgastonochoa commented on drivers/spi/spi.c:
```diff
@@ -0,0 +1,19 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "spi.h"
+
+bool spi_buf_set_in_nocache(const struct spi_buf_set *bufs)
+{
+	for (size_t i = 0; i < bufs->count; i++) {
+		const struct spi_buf *buf = &bufs->buffers[i];
+
+		if (!dt_buf_in_nocache((uintptr_t)buf->buf, buf->len)) {
+			return false;
+		}
+	}
+	return true;
+}
```
> can you make this static inline and move this function into include/zephyr/drivers/spi.h so to not have at all spi.c and spi.h from this PR?

It was static inline before and I moved it because it was requested in comments above, and I agreed as I prefer to have implementations in C files since tests are simpler.

teburd replied:
agree with @carlocaione this should be in spi.h as a static inline

dgastonochoa commented on include/zephyr/devicetree/mem.h:
> This is wrong on many levels. See the other comments but the point is that we must try to keep all this work contained in the spi layer.

That is what I had in my first commit about 2 months ago, before getting several requests to change it and make it generic. This is explained in the first and second descriptions so I won't dive into it anymore.

carlocaione replied:
but the problem is that this is not generic at all, you are only dealing with the nocache attribute. If you want to add generic helpers for the memory attributes you must do that supporting any kind of attribute, not just one, otherwise at least this is not scalable at all.

dgastonochoa commented on include/zephyr/devicetree/mem.h:
```diff
@@ -0,0 +1,107 @@
+ * a string that specifies such region's properties.
+ *
+ */
+struct dt_mem_region {
+	const uintptr_t start;
+	const uintptr_t end;
+	bool isnocache;
+};
```
Yes, I mentioned before I was doubtful about leaving the attributes string, but I didn't have a strong opinion on it so I wanted to receive some feedback.

teburd reviewed (commented):
I don't really understand the complaints here @carlocaione, the dt macros and types can easily be extended and encode at this moment the most useful information, the start, end, and cached attributes. This seems like it could be extended further as needed?

Or is the issue that its working on a struct rather than using DT macro wrappers for getting each attribute of a memory region?

Unclear to me what the problem is.

carlocaione commented:
@teburd 

> Unclear to me what the problem is.

There are several problems:
1. This is not scalable, at all. The current solution is implying that you have one check function per attribute. This should be generic, with the attribute as parameter.
2. Yes, this is incomplete. The problem is that as soon as you start thinking how to extend this and how to make it generic, you will soon realize that this approach is too simplistic (been there, done that).
3. `include/zephyr/devicetree/mem.h` is not the correct place for functions like `dt_buf_in_nocache`.  That function is not related to DT at all, in there you should only find DT macros. See for example the difference between `include/zephyr/devicetree/mbox.h` and `include/zephyr/drivers/mbox.h`.
4. The header file is using a const struct array built at compile-time. If anything, that array should be passed as parameter and the function only checking that.

I'm fine with the general idea of this PR, but this is not the correct (and complete) solution to a complicated problem. This problem cannot be solved with DT macros only or DT helpers, unfortunately something resembling a subsystem must be introduced and I'm working on that. If this work is kept in the realm of SPI (or platform specific even) I have no complains and this is what I'm suggesting here.

teburd commented:
> @teburd
> 
> > Unclear to me what the problem is.
> 
> There are several problems:
> 
> 1. This is not scalable, at all. The current solution is implying that you have one check function per attribute. This should be generic, with the attribute as parameter.

e.g. DT_MEM_HAS_ATTR(DT_NODE(some_mem_region), NOCACHE) kind of thing?

> 2. Yes, this is incomplete. The problem is that as soon as you start thinking how to extend this and how to make it generic, you will soon realize that this approach is too simplistic (been there, done that).

Broadly... simpler things tend to win out

> 3. `include/zephyr/devicetree/mem.h` is not the correct place for functions like `dt_buf_in_nocache`.  That function is not related to DT at all, in there you should only find DT macros. See for example the difference between `include/zephyr/devicetree/mbox.h` and `include/zephyr/drivers/mbox.h`.
> 4. The header file is using a const struct array built at compile-time. If anything, that array should be passed as parameter and the function only checking that.

Agree, should be static const in a .c and used with an extern in the header, no argument there.

> 
> I'm fine with the general idea of this PR, but this is not the correct (and complete) solution to a complicated problem. This problem cannot be solved with DT macros only or DT helpers, unfortunately something resembling a subsystem must be introduced and I'm working on that. If this work is kept in the realm of SPI (or platform specific even) I have no complains and this is what I'm suggesting here.

Understood, I guess I'll be interested to see what that looks like. Until then I think we should give the PR author a clear answer. It sounds like you are working on a more general proposal to solve this which is great. Given that I don't think the requirement to generalize here in this PR makes sense. So @dgastonochoa it seems like the simplest path forward is to revert back to what you had initially. Sorry about that! I know this idea is more broadly useful to other platforms (Intel and NXP have similar things) and so I was hopeful we could get the start of something going here. I guess @carlocaione will have something for us to peruse soon in that regard. Looking forward to it!

carlocaione commented:
> e.g. DT_MEM_HAS_ATTR(DT_NODE(some_mem_region), NOCACHE) kind of thing?

Yeah, but the parameters that you want the function / macro to take are address, size and attribute of the region you want to check. So this is not really a `DT_MEM_HAS_*` kind of thing but possibly a real function `mem_check_buf(void *addr, size_t size, int attr)` where the address is possibly a virtual address while the DT addresses are physical addresses, so you have also some MMU complexity in the mix.

This, together with your correct other comment:
> Agree, should be static const in a .c and used with an extern in the header, no argument there.

means (to me) that this is a problem that cannot be exclusively fixed at `include/zephyr/devicetree/` level, but this requires a proper library / subsystem that is basically gathering all the info coming from DT about memory regions and act on this information using a proper API.

I'm currently working on that, I will have something ready in the next few days.

dgastonochoa commented:
> So @dgastonochoa it seems like the simplest path forward is to revert back to what you had initially.

This will involve to move **all** the logic to `spi_ll_stm32.c`, nothing will be in headers, not even in `spi.h`, as it is done in other drivers. @carlocaione @teburd could you please confirm you are both happy with this?

teburd commented:
> > So @dgastonochoa it seems like the simplest path forward is to revert back to what you had initially.
> 
> This will involve to move **all** the logic to `spi_ll_stm32.c`, nothing will be in headers, not even in `spi.h`, as it is done in other drivers. Also, all logic will be restricted to be used only for stm32h7. @carlocaione @teburd could you please confirm you are both happy with this?

Yes that's fine

dgastonochoa commented:
@carlocaione @teburd all changes applied. Could you review them please?

carlocaione commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -51,6 +51,33 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
+		.start = (const uintptr_t)DT_REG_ADDR(node_id),                         \
+		.end =                                                                  \
+			(const uintptr_t)(DT_REG_ADDR(node_id) + DT_REG_SIZE(node_id)) - 1, \
+		.isnocache = IS_NOCACHE_MEM_REGION(node_id)                             \
```
if you are interested only in the `NOCACHE` regions, why not filtering out the other regions when constructing `mem_regions[]`?

dgastonochoa replied:
I think you mean using `COND_CODE_1` in combination with `IS_NOCACHE_MEM_REGION` to not add "no nocache" entries to the array? That would be ideal, I will try.

dgastonochoa replied:
Done

carlocaione commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -744,6 +799,13 @@ static int transceive_dma(const struct device *dev,
 	}
 
+#ifdef CONFIG_SOC_SERIES_STM32H7X
+	if ((tx_bufs != NULL && !spi_buf_set_in_nocache(tx_bufs)) ||
```
I think the condition `tx_bufs != NULL`  should be checked by `spi_buf_set_in_nocache` to do a bit of parameters sanitation.

dgastonochoa replied:
That makes things more confusing in my eyes. It would require `spi_buf_set_in_nocache` tu return an error code instead of a `bool` to identify when the buffers are, or are not, in nocache, and when the input parameter is an error. That's not the intention with that function, I want to to return a `bool` to be coherent and simple.

Also, it obscures the semantic of that `tx_bufs` and `rx_bufs` should only be checked if the SPI transaction involves read and/or write respectively. Checking parameters inside that function tells that `tx_bufs` need to be checked even when the SPI operation is read only, or that `rx_bufs` need to be checked when the operation is write only. I want to make it clear that they only need to be checked if the operations requires so.

Finally, a `NULL` is essentially address = 0x0. That technically is an address as valid as any other one to be checked to verify if it is nocache or not. Providing a special treatment to it confuses what the function does.

carlocaione replied:
fair enough

carlocaione commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -51,6 +51,37 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
+
+#ifdef CONFIG_SOC_SERIES_STM32H7X
+
+#define IS_NOCACHE_MEM_REGION(node_id)                                          \
+	COND_CODE_1(DT_ENUM_HAS_VALUE(node_id, zephyr_memory_attr, RAM_NOCACHE),    \
+			(1),                                                                \
+			(0))
```
A better way to do this: https://github.com/zephyrproject-rtos/zephyr/pull/60850

carlocaione commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -723,6 +754,34 @@ static int wait_dma_rx_tx_done(const struct device *dev)
+		const struct mem_region *mem_reg = &nocache_mem_regions[i];
+
+		const bool buf_within_bounds =
+			(buf >= mem_reg->start) && ((buf + len_bytes - 1) <= mem_reg->end);
```
why you do not remove the `-1` from `mem_reg->end` and just:
```c
(buf + len_bytes) < mem_reg->end
```

dgastonochoa replied:
I did this on purpose. In general terms, this is checking if a number X is within a closed interval. If you would need to do this, you would naturally use <= and >=. I wanted to follow this pattern as it seems more intuitive. Otherwise, the bool variable is called "within_bounds" but it seems to not be doing exactly that. It seams more clear to me this way. In other contexts (most of them possibly), I would do as you say.

carlocaione commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -51,6 +51,37 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
+
+#define GET_MEM_REGION(node_id)                                                 \
+	{                                                                           \
+		.start = (const uintptr_t)DT_REG_ADDR(node_id),                         \
```
I don't think this cast is needed.

dgastonochoa replied:
Done

carlocaione reviewed (commented):
Just a few nits but we are getting there
